plugins/formal-verification/formver.core/src/org/jetbrains/kotlin/formver/conversion/StmtConversionVisitor.kt

```diff
@@ -55,23 +55,19 @@ class StmtConversionVisitor : FirVisitor<Exp, StmtConversionContext>() {
         val expr = returnExpression.result.accept(this, data)
         // TODO: respect return-based control flow
         val returnVar = data.signature.returnVar
-        data.addStatement(Stmt.LocalVarAssign(returnVar.toLocalVar(), expr))
+        data.addStatement(Stmt.LocalVarAssign(returnVar.toLocalVar(), expr.withType(returnVar.type.type)))
         return UnitDomain.element
     }
 
     override fun visitBlock(block: FirBlock, data: StmtConversionContext): Exp =
         // We ignore the accumulator: we just want to get the result of the last expression.
-        block.statements.fold(UnitDomain.element) { _, it -> it.accept(this, data) }
+        block.statements.fold<FirStatement, Exp>(UnitDomain.element) { _, it -> it.accept(this, data) }
 
     override fun <T> visitConstExpression(constExpression: FirConstExpression<T>, data: StmtConversionContext): Exp =
         when (constExpression.kind) {
             ConstantValueKind.Int -> Exp.IntLit((constExpression.value as Long).toInt())
             ConstantValueKind.Boolean -> Exp.BoolLit(constExpression.value as Boolean)
-            /* TODO: For now null is always hard-coded to be of type Nullable[Int].
-             * This needs to be generalized and for this the type of the return expressions needs to be known.
-             * This type should maybe be passed as function argument.
-             */
-            ConstantValueKind.Null -> NullableDomain.nullVal(Type.Int)
+            ConstantValueKind.Null -> NullableDomain.nullVal((data.embedType(constExpression.typeRef.coneType) as NullableTypeEmbedding).elementType.type)
```

Comment:
Is it worth making a helper function here, something like `embedType` branching on nullability and calling either `embedNullableType` or `embedNonNullableType`?

Reply:
I don't know if it's worth having 3 functions for this, especially because you have to make extra logic to handle the cases when someone passes a nullable type to `embedNonNullableType` or a non-nullable type to `emdedNullableType`.